### PR TITLE
New semantic tests extracted

### PR DESCRIPTION
**Pull request by anurag-git**
### Description

Extracted and added semantic tests for
1. Transaction Status
2. Empty Contract
3. Smoke test for Range

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### CODE REVIEW

**Comment by anurag-git** [comment]
Partial fixes #4223

**Comment by leonardoalt** [comment]
There are some trailing whitespaces (please check the failing test)

**Comment by leonardoalt** [comment]
Where were these tests extracted from? Can they be removed from the cpp test file(s) accordingly?

**Comment by anurag-git** [comment]
> There are some trailing whitespaces (please check the failing test)

Done

**Comment by anurag-git** [comment]
> Where were these tests extracted from? Can they be removed from the cpp test file(s) accordingly?

@leonardoalt ,   This has been taken from SolidityEndToEndTest.cpp. I am not sure, if I can just delete them.
Also, any test to check the sanity of above file after removing the tests.

**Comment by ekpyron** [comment]
@anurag-git The main purpose of extracting the tests is to be able to remove them from ``SolidityEndToEndTest.cpp``, so yes: please go ahead and do that - it'll also make it easier to review the extraction in the PR.

Regarding the sanity of ``SolidityEndToEndTest.cpp``: If you remove an entire ``BOOST_AUTO_TEST_CASE(...) { ... }`` block, then you're fine. If it's only parts of the tests you can check, whether it still compiles and run ``build/test/tools/soltest`` with an instance of ``aleth`` running in the background to execute the end-to-end-tests, but I don't expect problems there and we'll see any issues in the CI tests running on the PR anyways.

**Comment by anurag-git** [comment]
> @anurag-git The main purpose of extracting the tests is to be able to remove them from `SolidityEndToEndTest.cpp`, so yes: please go ahead and do that - it'll also make it easier to review the extraction in the PR.
> 
> Regarding the sanity of `SolidityEndToEndTest.cpp`: If you remove an entire `BOOST_AUTO_TEST_CASE(...) { ... }` block, then you're fine. If it's only parts of the tests you can check, whether it still compiles and run `build/test/tools/soltest` with an instance of `aleth` running in the background to execute the end-to-end-tests, but I don't expect problems there and we'll see any issues in the CI tests running on the PR anyways.

Thanks, changes done and pushed.

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 1]
```diff
@@ -64,47 +64,6 @@ int constexpr roundTo32(int _num)
...
-	compileAndRun(sourceCode);
-	testContractAgainstCppOnRange("f(uint256)", [](u256 const& a) -> u256 { return a * 7; }, 0, 100);
-}
-
-BOOST_AUTO_TEST_CASE(empty_contract)
-{
-	char const* sourceCode = R"(
-		contract test { }
-	)";
-	ALSO_VIA_YUL(
```
Comment (ekpyron) [inline test/libsolidity/SolidityEndToEndTest.cpp:102]: We have to be careful here - if you just extract tests using the ``ALSO_VIA_YUL`` macro, then they will only be compiled without yul and not "also via yul" :-). There's a "compileViaYul" option for the extracted tests, but that then *only* compiles via Yul and not *also*... we're still discussing how best to deal with this and we'll get back to you.

**Comment by leonardoalt** [comment]
Discussion from Gitter: this PR depends on the yet non-existing feature `alsoViaYul` that should run both with/without for the extracted tests

**Comment by ekpyron** [comment]
@anurag-git It might take a bit of time until we have a good solution for the ``ALSO_VIA_YUL`` part.

If you want to go ahead with this soon, there are two options: either you just *don't* extract tests that involve ``ALSO_VIA_YUL`` or you create *two* files for each of these tests, one regular version and one version suffixed "_via_yul", which sets
```
// ====
// compileViaYul: true
// ----
// <test expectations>
```
in the expectations (you can find examples of such tests in ``test/libsolidity/semanticsTests/viaYul/*``).

**Comment by anurag-git** [comment]
> @anurag-git It might take a bit of time until we have a good solution for the `ALSO_VIA_YUL` part.
> 
> If you want to go ahead with this soon, there are two options: either you just _don't_ extract tests that involve `ALSO_VIA_YUL` or you create _two_ files for each of these tests, one regular version and one version suffixed "_via_yul", which sets
> 
> ```
> // ====
> // compileViaYul: true
> // ----
> // <test expectations>
> ```
> 
> in the expectations (you can find examples of such tests in `test/libsolidity/semanticsTests/viaYul/*`).

Took 2nd approach and pushed.

**Comment by leonardoalt** [comment]
I'm still not sure about this solution with 2 files.
Are we going to merge them back into a single file when we have the `alsoViaYul` option?

**Comment by anurag-git** [comment]
> I'm still not sure about this solution with 2 files.
> Are we going to merge them back into a single file when we have the `alsoViaYul` option?

@ekpyron , Would need your help here? I am ok with both options.

**Comment by ekpyron** [comment]
I think I'll create a way to run both kinds of tests at once after all - an imperfect, but working version will be ready shortly - I'll let you know when it's done, then you can continue based on that (sorry for the resulting delay).

**Comment by ekpyron** [comment]
@anurag-git We just merged #6916, so if you rebase this PR on top of the most current ``develop``, you'll be able to use ``compileViaYul: also`` for all the tests that have ``ALSO_VIA_YUL`` in ``SolidityEndToEndTests.cpp`` and we don't need two files, but only one. Sorry for the back and forth in this matter :-)!

**Comment by anurag-git** [comment]
> @anurag-git We just merged #6916, so if you rebase this PR on top of the most current `develop`, you'll be able to use `compileViaYul: also` for all the tests that have `ALSO_VIA_YUL` in `SolidityEndToEndTests.cpp` and we don't need two files, but only one. Sorry for the back and forth in this matter :-)!

I hope the changes are correct. Finally.

**Review by ekpyron — DISMISSED** [review]

**Comment by ekpyron** [comment]
@anurag-git Thank you very much! One final thing before we can merge this: can you squash the PR into one single commit?

File: test/libsolidity/SolidityEndToEndTest.cpp [hunk 2]
```diff
@@ -64,47 +64,6 @@ int constexpr roundTo32(int _num)
...
-
-BOOST_AUTO_TEST_CASE(smoke_test)
-{
-	char const* sourceCode = R"(
-		contract test {
-			function f(uint a) public returns(uint d) { return a * 7; }
-		}
-	)";
-	compileAndRun(sourceCode);
-	testContractAgainstCppOnRange("f(uint256)", [](u256 const& a) -> u256 { return a * 7; }, 0, 100);
```
Comment (ekpyron) [inline test/libsolidity/SolidityEndToEndTest.cpp:94]: By the way: for this particular test here, it's not a problem and I don't really mind, but this command *actually* tests every value between ``0`` and ``100``, whereas the extracted test only checks two values, namely ``0`` and ``99``.

So actually tests involving ``testContractAgainstCppOnRange`` should probably stay in ``SolidityEndToEndTest.cpp``.

But yeah, in this particular case, it really doesn't matter, if the full range is checked or not, so it's fine.

**Review by ekpyron — DISMISSED** [review]
I cleanup up the damage this took during rebasing. Now it's fine to be merged, once the tests pass again!

**Review by ekpyron — APPROVED** [review]

File: test/libsolidity/SolidityEndToEndTest.cpp (hunk 2 above)
Comment (anurag-git) [reply test/libsolidity/SolidityEndToEndTest.cpp:94]: Ok. I will keep in my mind for future changes

**Pull request merged**